pallets/gear-messenger/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use common::storage::{
+        Callback as GearCallback, DeckError as GearDeckError, Messenger as GearMessenger, NextKey,
+        Node, StorageCounter as GearStorageCounter, StorageDeck as GearStorageDeck,
+        StorageFlag as GearStorageFlag, StorageMap as GearStorageMap,
+        StorageValue as GearStorageValue,
+    };
+    use frame_support::{pallet_prelude::*, traits::ConstBool};
+    use frame_system::pallet_prelude::*;
+    use gear_core::message::StoredDispatch;
+    use scale_info::TypeInfo;
+    use sp_std::{convert::TryInto, marker::PhantomData, prelude::*};
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>>
+            + IsType<<Self as frame_system::Config>::Event>
+            + TryInto<Event<Self>>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    pub enum Event<T> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
```

Comment:
Do you mean to copy-paste DequeError here?
There will be other errors related to mailbox, WL, etc

Reply:
I would also like to add variable logging in these algos for debugging

Reply:
I think in order to avoid copy-paste of DequeError and to make pallet errors more precise, we can make a generic Error type in `StorageDeque` trait with some trait bound like `DequeError` to be implemented by `pallet_messenger::Error`. What do you think?
Actually if it's an over-featuring, then no need for that. I mean reaching precise pallet errors should not be so costly.

Reply:
I think it's over-engineering for that time. We may run into many troubles and more code len with such method. But will consider some time soon.